File: inventory3.py
# Browser Simulation using Stack

history = []          # Stack for visited pages
forward_stack = []    # Stack for forward navigation


def visit_page(page):
    history.append(page)   # Push onto history
    forward_stack.clear()  # Clear forward stack on new visit
    print(f"Visited: {page}")


def go_back():
    if not history:
        print("No pages in history.")
        return
    last_page = history.pop()   # Remove current page
    forward_stack.append(last_page)
    if history:
        print(f"Going back... Current page: {history[-1]}")
    else:
        print("No pages left in history.")


def go_forward():
    if not forward_stack:
        print("No forward pages available.")
        return
    page = forward_stack.pop()
    history.append(page)
    print(f"Forward to: {page}")


def show_history():
    if not history:
        print("History is empty.")
    else:
        print("History:", " -> ".join(history))


# ---------------- Menu ----------------
while True:
    print("\n1. Visit Page")
    print("2. Back")
    print("3. Forward")
    print("4. Show History")
    print("5. Exit")

    choice = input("Enter choice: ")

    if choice == "1":
        page = input("Enter page name: ")
        visit_page(page)
    elif choice == "2":
        go_back()
    elif choice == "3":
        go_forward()
    elif choice == "4":
        show_history()
    elif choice == "5":
        print("Exiting browser simulation.")
        break
    else:
        print("Invalid choice.")

File: Inventory3.cpp
#include <iostream>
#include <stack>
#include <vector>
using namespace std;

class Browser {
    stack<string> history; // stack to store pages in reverse
    vector<string> order;  // to keep order for history display

public:
    // Visit a new page
    void visitPage(const string& page) {
        history.push(page);
        order.push_back(page);
        cout << "Visited: " << page << endl;
    }

    // Go back to previous page
    void goBack() {
        if (history.empty()) {
            cout << "No pages in history." << endl;
            return;
        }
        string lastPage = history.top();
        history.pop();
        order.pop_back();

        cout << "Going back from: " << lastPage << endl;
        if (!history.empty()) {
            cout << "Current page: " << history.top() << endl;
        } else {
            cout << "No pages left in history." << endl;
        }
    }

    // Show browsing history
    void showHistory() {
        if (order.empty()) {
            cout << "History is empty." << endl;
            return;
        }
        cout << "History: ";
        for (size_t i = 0; i < order.size(); i++) {
            cout << order[i];
            if (i != order.size() - 1) cout << " -> ";
        }
        cout << endl;
    }
};

int main() {
    Browser browser;
    int choice;
    string page;

    while (true) {
        cout << "\n1. Visit Page" << endl;
        cout << "2. Back" << endl;
        cout << "3. Show History" << endl;
        cout << "4. Exit" << endl;
        cout << "Enter choice: ";
        cin >> choice;

        switch (choice) {
            case 1:
                cout << "Enter page name: ";
                cin >> page;
                browser.visitPage(page);
                break;
            case 2:
                browser.goBack();
                break;
            case 3:
                browser.showHistory();
                break;
            case 4:
                cout << "Exiting browser simulation." << endl;
                return 0;
            default:
                cout << "Invalid choice." << endl;
        }
    }
}
